Codepen links

paddymul opened the pull request:
- [ ] issues: fixes #8692
- [ ] tests added / passed
- [ ] release document entry (if new feature or API change)

paddymul commented:
A couple of issues.

1. I linked to my personal codepen site.  This could be a security concern.  Is there an official Bokeh jsfiddle or codepen account?
2. There are broken codepen examples, these are waiting on @mattpap to fix #8686 and #8597

bryevdv commented:
There is a JSFiddle (currently disused) but not a Codepen, though we could certainly create a codepen. I'd prefer a solution that allowed managing the contents of the examples from a GH repo or similar. Keeping the code up to date was a huge hassle when we used to have examples on JSFiddle. Speaking of which, all of these would need to be updated to use 1.1.0, version 0.12.5 is nearly two years old at this point and I am 100% not interested in demonstrating/supporting anything from the past (I simply don't have the bandwidth).

paddymul commented:
I understand your concerns about maintenance.  I was just putting code into the codepen directly from the documentation, so that's why it's pointing at 0.12.5

I'm not sure what the best way to handle the js versioning in the documentation is.  Is there a way to create a latest js file on each release so in addition to https://cdn.pydata.org/bokeh/release/bokeh-0.12.5.min.js there could be a https://cdn.pydata.org/bokeh/release/bokeh-latest.min.js ?  That or is there a release template variable in sphinx?  I could template the code blocks to always point at $RELEASE_NUMBER for the SCRIPT tags.

I think I can come up with a solution for directly populating Codepens from the PRE block in the documentation using [Codepen Prefill](https://blog.codepen.io/documentation/api/prefill/) and some javascript in the documentation.   The other nice thing about using the prefill API is that it seems to create a new anonymous codepen, I think this would be secure (as opposed to linking to a paddy_mullen owned codepen where I could change the code without the URL changing).  

I'm not tied to Codepen, but when searching for some solution for this code synchronization problem (github, api,...) amongst the various JSFiddle like services (JSFiddle, Codepen, jsbin...) this seems like the most workable solution.  I'm open to suggestions.  

@bryevdv would you be ammenable to inline JS + Codepen to solve this?

bryevdv commented:
I'm definitely amenable to inline JS in the docs, I think that is preferable from a maintenance standpoint

paddymul commented:
Ok, I integrated the codepen default JS injection link code.

Here is what the links end up looking like

![screen shot 2019-03-01 at 9 10 40 am](https://user-images.githubusercontent.com/40453/53650675-495ddc80-3c02-11e9-97c0-f72767d2ac9d.png)

A couple of things.

1. I split the bottom combined example into an HTML block and a javascript block.  The HTML block is used as the template for all additional codepen links.  This also results in most elements having a "Add to plot" link.  It might be possible to not use a whole HTML block by using the css_external and js_external properties on the prefill call.
2. Every codepen has the same title, currently defaulting to "Cool Pen".  I can't currently think of a dependable way to extract a title for each codepen without inserting additional markup into the sphinx document.
3. What are your thoughts for the embedded styles?  I'm open to better styling.  It would be relatively easy to add a raw::html block after each example to include a button element.
4. The first javascript block gets a codepen link too.  "var xdr = new Bokeh.Range1".

bryevdv commented:
@paddymul I'd forgotten we only have static images in the docs at the moment. I guess whatI'd really like is:

> a single source of truth that drives an inline live plot in the docs, and that can be used to pre-fill a codepen that users can launch

I think given the amount of JS/HTML mucking that is necessary to prefill a codepen, that this might call out for a sphinx extension, i.e. to be able to do:

```
.. bokehjs-plot:: path/to/script.js
    :prefill:
```
Then this would generate a live plot in the docs from `script.js`, and also (when `:prefill:` is present) pre-fill a codepen provide the link to it. 


Something like this would make it easy to expand the BokehJS docs with lots of examples over time, and to maintain the code for the examples in one place under revision control. I can definitely help with the sphinx extension bits, but I would need help with all the JS bits. I'd propose to invite you to @bokeh/dev so that we could easily collaborate on a single branch if that's ok? If so, I can get a PR with a scaffold sphinx extension set up and ping you to fill in the JS/HTML bits. Does that sound like a reasonable plan?

paddymul commented:
Creating a sphinx extension sounds awesome.  

I think the extension should take the following args (python style) if possible.
```
def js_codeblock(path_to_script, title, html_template=None, css_template=None):
    pass
```
The javascript for the add to codepen really only needs to be written once and added to the whole page, I don't think there will need to be any special javascript on each code block other than a special class name.  

I think it's easier for the codepen side to have separate code blocks for JS and HTML, this is a presentation question though and I defer to the team.

I would want the extension to output html that looks aproximately like this
```
<div class="codepen-block" data-title="Scatter Example"
        data-js='["bokeh.$CURRENT_VERSION.min.js", "bokeh-api.$CURRENT_VERSION.min.js"]'>
    <div class="highlight-javascript notranslate">
        //js-code
    </div>
    <div class="highlight-html notranslate">
        <!--  this block is optional -->
    </div>
    <button class="open-in-codepen">Open in codepen</button>
</div>
```

A couple of questions for  @bokeh/dev 
1. Who is responsible for styling guidelines for the docs.  What should the codepen link look like?  How should we handle the styling process?
2. How many pages do you expect to have codepen js links on?

ps:
I'm finding the tooling around restructured text really touchy.  Every time I make a change to an rst file I have to wait about 1.5 to 5 minutes for all of the documentation to rebuild frequently after all of that I find out that I have introduced a syntax error.

paddymul commented:
Where does this stand?  What can I contribute to move this PR along?  I'm happy to take a stab at the sphinx extension, but I would like some feedback on the docs aesthetics.

bryevdv commented:
@paddymul I've gotten sidetracked, sorry. 

> Who is responsible for styling guidelines for the docs. What should the codepen link look like? How should we handle the styling process?

No one currently, so anything reasonable seems acceptable

> How many pages do you expect to have codepen js links on?

Just the one BokehJS user guide page, unless that page gets split up

> I'm finding the tooling around restructured text really touchy. Every time I make a change to an rst file I have to wait about 1.5 to 5 minutes for all of the documentation to rebuild frequently after all of that I find out that I have introduced a syntax error.

Are you always issuing `make clean` ? If you just call `make html` then it should only rebuild what necessary (except after an error --- everything is built from scratch after an error)

I'm happy to have you work on the sphinx extension and just advise :) but I should warn that if you are new to sphinx extensions there is a bit of a learning curve. There are a variety of ways this extension might be made to work, but I think given how much HTML/JS work there is to do, it could just follow the patter of the "todo" tutorial pretty closely:

https://www.sphinx-doc.org/en/stable/extdev/tutorial.html

That is, define a new node type to represent `.. bokehjs-plot::` and have the reading stage just do almost nothing except insert these nodes. Then all the work is done by an html visitor in the writing stage (it just emits all the JS and HTML you want, in this case, the codeine block). If we wanted to start simple we could have *just* the codeine block but obviously I think embedding a real live plot would be great (definitely possible, just more work)

paddymul commented:
I added a bokehjs-block directive, it's based on the collapsible_code_block directive.  There is now a single place to edit the js version  [bokehjs_block_epilogue](https://github.com/paddymul/bokeh/blob/codepen_links/bokeh/sphinxext/_templates/bokehjs_block_epilogue.html) .  This is still not ideal, because the version number isn't templatized, but it's a step in the right direction. 

You have mentioned that the bokehJS docs having images is surprising.  Did you want live plots?   Can we handle that in a separate issue/pull request?

Thoughts?

bryevdv commented:
@paddymul following CCB seems reasonable, I think I'd prefer the name `bokehjs-content` or similar rather than "block" in this case. We'll definitely need to inject the real version into the template at docs generation time. You can see how the plot directive handles the "docs CDN" link here:

https://github.com/bokeh/bokeh/blob/master/bokeh/sphinxext/bokeh_plot.py#L268

I suppose we could forego adding real live plots for now,  though I'd certainly prefer to finish thing in one PR. Is there any dead-simple option? I.e. is it possible to IFrame in the output cell from the codepen some directly way? 

Also, is all the updating necessary in layout.html? Couldn't all that be done in the HTML prologue/epilogue for the BokehJS content? Would like to keep the implementation details for this self-contained. Coupling the details of `bokehjs-content` to the general template is a maintenance headache.

paddymul commented:
@bryevdv Thanks for your patience on this.

1. bokehjs-content sounds better, I'll make that change.

2. For templating the js files, I took a look at the resources bit from bokeh_plot.py  , it looks like I want to copy the same `resources` logic into bokehjs-content and use resources.js_files and resources.css_files to inject into the template.  Does that sound right?

3. 

The reason I changed layout.html was that there is a bit of global js and HTML that is needed for this integration.  I initially injected this same js and html into each bokehjs-content block, but that seemed ugly to me.  The JS basically sets an onload handler, which first checks for a global variable to see if this function had already been called, if so it exits, otherwise it adds the proper onclick function to each "Try on Codepen button".  This struck me as ugly and it would make the markup for the bokehjs-content block harder to be reliably edited.  

I see a couple of alternative implementations.
a. Insert the embedded HTML and JS onload handler into each bokehjs-content directive.
b. Create a separate bokehjs-preamble directive that inserts the boilerplate onto the page.  This would need to be included in the RST file in addition to the bokehjs-content directives.  I would like to couple this with a failsafe check in each bokehjs-content directive that uses js to check to see that the bokehjs-preamble codeblock was inserted into the page, if not it would call `alert('you must use the bokehjs-preamble directive for the codepen links to properly function');`.  The idea being that this makes it easier for future maintainers to understand why the JS isn't working.
c. If it is possible for sphinx to inject special content on the first call to a directive on a page, only inject the preamble on the first call on a page to bokehjs-content.  I don't know enough about sphinx to figure this out, but if someone provides a skeleton, I'm happy to flesh out the implementation.  This would be the cleanest solution.

4. I guess I just want to cognitively wrap up this sphinx - codepen integration before adding more features.  Once I have a working implementation that we are all happy on for just the sphinx extensions and codepen integration, making the plots live should be fairly straightforward.  

5.  When is the next bokeh release?  I really want to do the work on my side to get my pull request with at least codepen links and preferably codepen links + live bokehjs plots merged into the next release.

bryevdv commented:
@paddymul 

2) sure, or if it makes sense, factoring common logic out to `sphinxext/utils.py` would also be welcome

3) The way you can inject content once is to use the `html-page-context` event:

    https://www.sphinx-doc.org/en/1.8/extdev/appapi.html#sphinx-core-events

  You could then add just a template parameter to `layout.html`. Then if `bokeh-content` is invoked on a page, it records this page for later (see the `todolist` sphinx tutorial for a similar example). Then the callback for `html-page-context` conditions on whether the current page previously had `bokeh-content` invoked --- if it did, the callback injects your special content for the template parameter. Otherwise, it does nothing / injects the empty string. Let me know if that is enough detail 

5) hopefully the next week or two. There is one last layout related issue to sort out.

paddymul commented:
Ok, I think I'm making progress.

I created the bokehjs-preamble directive.  This templatizes the js and css for sending to codepen.  The HTML template is inserted into an HTML comment on the docs side, the javascript extracts the text of this HTML comment to send it to CodePen.  It's a little hacky, but this method will keep the HTML of the documentation clean.

I needed to change resources.py to add "extra_js_components", this allows the inclusion of `bokeh-api`.  The documentation is the only place I think it would really be necessary to have python output `bokeh-api` js files.  If you think this usecase warrants another mode, I can make that change, but I don't think it's necessary.

I'm going to look at making the plots live embedable from the js code now.

There are also codepens that don't work, specifically the last codepen on this page because @mattpap changes from  #8686 don't have a version and corresponding artifact on the CDN.  I think those codepens will work, but I'm not certain.

There are other references in the docs to the bokehjs API changing with 1.0 that I will remove in this PR or a separate PR.

bryevdv commented:
FYI @paddymul our Appveyor jobs are at a premium, please consolidate many commits into single or fewer commits when possible

> I created the bokehjs-preamble directive. This templatizes the js and css for sending to codepen.

Maybe let's have a name reflect that, e.g. `bokehjs-codepen-init` FWIW I had imagined handling this "under the covers" i.e. by setting an internal flag that indicated a BokehJS resource load was needed for a page, and having the html template event for the page condition on that flag. I'm just very reticent to add new public API nowadays, since it's basically forever. 

> I needed to change resources.py to add "extra_js_components"

If we can avoid adding this, I think we should. It seems reasonable that `bokeh-api` should be handled with the existing `components` parameter like everything else?

> I'm going to look at making the plots live embedable from the js code now.

I'm happy to punt on this for now.

paddymul commented:
@bryevdv I think I have addressed all of your points.  

My understanding is that the CI jobs are only triggered on the last commit pushed, so even though I pushed a bunch of commits here, I don't expect to trigger multiple appveyor jobs.  Let me know if I'm mistaken.

Also let me know if the way I'm using `Resources` and running `resources.js_components.append('bokeh-api')` was what you had in mind.

I think setting a flag on the source document for only once inclusion is much cleaner than trying to use sphinx events.

bryevdv commented on bokeh/sphinxext/bokehjs_content.py:
```diff
@@ -0,0 +1,187 @@
+    # Otherwise assume it is a dev/rc/full release version and use CDN for it
+    else:
+        resources = Resources(mode="cdn", version=docs_cdn)
+resources.js_components.append("bokeh-api")
```
I think most of this can be removed, since the codepen will not function with anything other than CDN resources. I think you just need:
```
resources = Resources(mode="cdn")
resources.js_components.append("bokeh-api")
```
Things still won't work on CDN with e.g. local resources, but they wouldn't anyway, so we might as well keep this simple.

bryevdv replied:
actually, `BOKEH_DOCS_CDN=<real version>` can work too, so then:
```
docs_cdn = settings.docs_cdn()
if docs_cdn is None:
    resources = Resources(mode="cdn")
else:
    resources = Resources(mode="cdn", version=docs_cdn)
resources.js_components.append("bokeh-api")
```

bryevdv replied:
Ah, except that will fail with `BOKEH_DOCS_CDN="local"` since a version validation will throw an exception. OK, so then let's keep the logic as is, but not duplicate in two places. This shoudl be factored out into local `util.py ` or something

bryevdv commented:
@paddymul I checked this out locally and things look good. I left one comment regarding the resources. I think can (should) be simplified since the pen will never work with anything except real CDN resources (so don't bother with anything else like local or test resources)

re: the sphinx events, I think using the sphinx event would cleaner since it saves you from having to check whether anything has been seen before yourself. I wouldn't hold up the PR for that, though. But I might change it later, since having a mechanism to inject BokehJS on pages might be useful in a more general context later some day (thinking of re-working `:bokeh-plot:` to use JSON item embeds)

paddymul commented:
I made a separate `bokeh.util.sphinx` module with a `get_sphinx_resources` function.  This is used by `bokehjs_content` and `bokeh_plot`.  

A couple of notes - all of these relate to future development and I don't think they should hold up this PR:
1. If you do create the HTML event, the code I wrote doesn't actually inject bokehjs into the docs page, it only injects a short segment of JS that strips all formatting from code blocks and sends that along with HTML that points at the correct JS version over to codepen.
2. If a developer wants to play with experimental bokehjs libraries on codepen, that may be possible with a web debugging proxy like fiddler or charles.  I'm not sure if this would require a tweak to the resources config.

bryevdv commented:
@paddymul sorry for one last ask, I think this function is so specific and tied to `bokeh.sphinxext` that it should be `bokeh.sphinxext.util` instead of `bokeh.util`. It's not actually unimaginable that this gets pulled into a separate package in the future so it should all be kept together.

paddymul commented:
No problem. I thought of putting the module there for a second. It’s always the other door. I’ll get to it later today or tomorrow. 

Sent from my iPhone

> On Mar 17, 2019, at 8:27 AM, Bryan Van de Ven <notifications@github.com> wrote:
> 
> @paddymul sorry for one last ask, I think this function is so specific and tied to bokeh.sphinxext that it should be bokeh.sphinxext.util instead of bokeh.util. It's not actually unimaginable that this gets pulled into a separate package in the future so it should all be kept together.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

paddymul commented:
Changed

bryevdv commented:
@paddymul sorry I have to ask for one thing more. The last example should not be `bokehjs-content` That's because, as it states:

> A minimal example follows, *demonstrating a proper import of the libraries*, and dynamic creation and modification of plots.

But you've removed all the script tags that load BohekJS to make it work with the codepen extension. I think it would be nice to have this in codepen, but it's critical that we show users a *complete* example.

paddymul commented:
I agree. 

I think the best route is to add an `include_html` option to the bokehjs-content directive.  This is needed so that we get the templatized js/css files from `resources`.  The example in the docs was way out of date and needed to be manually editted with each new release.  I'll need to figure out how to send an `include_html` option bokehjs-content to codepen.  

What do you think of the other change I made to this example where instead of including the button in markup, I add the button to the DOM and attach the event handler from JS?  This makes for a slightly less clear example, but much easier templating.

Let me play with this a bit.  This type of feature also plays well with reading JS/HTML from a file.  I was initially hesitant to do this, but now I'm feeling more confident with the sphinx ecosystem.

paddymul commented:
The final complete example is now emitted as HTML with templated js libraries.  I added three options to `bokehjs-content` . 

```
js_file : string
    location of javascript source file
include_html: string
    if present, this code block will be emitted as a complete HTML template with
    js inside a script block
disable_codepen: string
    if present, this code block will not have a 'try on codepen' button.  Currently
    necessary when 'include_html' is turned on.
```

This is a very good foundation for `bokehjs-content`.  There is now a reference to the actual js code on the python side, this will help with live plot embedding.  

I had to copy a lot of `sphinx.directives.code.CodeBlock.run` and `bokeh.sphinxext.bokeh_plot.run` .  I didn't see straight forward ways to reuse the existing code because of references to `self`.


After this is merged I want to add two features to `bokehjs-content`

The next thing I want to work on is escaping the js and html, then injecting them into `data-js-source` attributes on `<div class="codepen-wrap">`.  The JS will then just read the `data-js-source` attribute, instead of parsing the actual code block.  I will do the same thing for a `data-html-attribute`, that will enable making the final complete example executable.  This will also simplify the injected JS on the page and move the escaping to the python side, that should make maintenance easier for the bokeh team.

Finally I can add a `body-html-file` option to the `bokehjs-content` directive so that extra html could be injected into the template.  This would allow putting actual `button` tags into the complete example.

I'm open to feedback on naming and implementation.

paddymul commented:
bokehjs-content now embeds live plots!!!  That was a shockingly easy change.

A couple of issues that I would like guidance on.
1.  I changed all code examples to use the following embedding form:
```    var doc = new Bokeh.Document();
    doc.add_root(p);
    Bokeh.embed.add_document_standalone(doc, document.currentScript.parentElement);
``` 
 Previously some examples used ```plt.show(p);```.  ```plt.show(p)``` inserts the plot at the bottom of the page which is not the desired behaviour.  I understand wanting to show some variety in the JS API, but I'm not sure how to proceed.
2. The charts.bar example (second to last `bokehjs-content` block) doesn't work.  I'll dig into this separately.  Once this example works it will make sense to remove the `image` directive.
3. The pie charts example is huge, I need to look up the api calls to make this render smaller.  Once it renders in a similar size to the existing `image` directive, I will remove the `image` directive.
4. The final example doesn't work.  This is a known bug.
5. The "Add Data" button for the final example displays in pink, just like the `Try on CodePen` buttons.  I think these should be styled differently.  Maybe the whole live embedded plot should be in a bordered `<div>` with a grayed background and an `H2`/`H3` saying "Live BokehJS Example" or something similar.  A button sitting in that grayed background would obviously be attached to the demo.  I'm open to suggestions.

bryevdv commented:
> What do you think of the other change I made to this example where instead of including the button in markup, I add the button to the DOM and attach the event handler from JS? This makes for a slightly less clear example, but much easier templating. 

I did not like it, to be honest. It seems to be adapting the example out of common usage in order to satisfy some unusual requirement of the tooling, and in general I think that is a bad thing. I would say the same thing about switching to `Bokeh.embed.add_document_standalone` usage. Normal usage is `show(plot)` so we should do whatever is necessary internally to demonstrate `show(plot)`. As an example the`:bokeh-plot` directive does some monkey-patching to make normal usage work in the context of Sphinx. I'm not sure what that translates to in this context. 

2-4) All sounds good

5)  I'm not super concerned about the button appearance, mostly that it needs to work.

paddymul commented:
`Bokeh.Plotting.show` is now monkey patched to insert the plot below the current script.  I included one example with `Bokeh.embed.add_document_standalone` as per the original docs.

bryevdv commented:
OK @paddymul I might want to take a look at this and change some things at some point, but it seems good for now once you figure out 2) and 3)

paddymul commented:
I fixed 2 and 3.

For 3, I changed some city names to abbreviations for cleaner display
![Screen Shot 2019-03-20 at 10 48 32 AM](https://user-images.githubusercontent.com/40453/54703684-db208180-4afe-11e9-9e1b-e9407bc86767.png)
to
![Screen Shot 2019-03-20 at 10 47 47 AM](https://user-images.githubusercontent.com/40453/54703696-e2478f80-4afe-11e9-8d8c-273068dc1ad7.png)

For issue 4
I tested that the example does work with the latest bokeh build from master.  I manually built bokehjs and uploaded it to my own s3 bucket.   https://codepen.io/paddy_mullen/pen/QoZGrZ?editors=1000  

I expect this to work properly when the CDN build is bumped.

I think this addresses all issues with this PR.

bryevdv commented:
OK let's go ahead and get this in and anything further we can do in follow-on work.
